pkg/yurthub/cachemanager/cache_manager.go

```diff
@@ -124,90 +126,183 @@ func (cm *cacheManager) QueryCache(req *http.Request) (runtime.Object, error) {
 func (cm *cacheManager) queryListObject(req *http.Request) (runtime.Object, error) {
 	ctx := req.Context()
-	comp, _ := util.ClientComponentFrom(ctx)
 	info, _ := apirequest.RequestInfoFrom(ctx)
-
-	key, err := util.KeyFunc(comp, info.Resource, info.Namespace, info.Name)
+	comp, _ := util.ClientComponentFrom(ctx)
+	key, err := cm.storage.KeyFunc(storage.KeyBuildInfo{
+		Component: comp,
+		Namespace: info.Namespace,
+		Name:      info.Name,
+		Resources: info.Resource,
+		Group:     info.APIGroup,
+		Version:   info.APIVersion,
+	})
 	if err != nil {
 		return nil, err
 	}
 
-	var gvk schema.GroupVersionKind
-	var kind string
-	// If the GVR information is not recognized, return 404 not found directly
-	gvr := schema.GroupVersionResource{
+	listGvk, err := cm.prepareGvkForListObj(schema.GroupVersionResource{
 		Group:    info.APIGroup,
 		Version:  info.APIVersion,
 		Resource: info.Resource,
+	})
+	if err != nil {
+		klog.Errorf("failed to get gvk for ListObject for req: %s, %v", util.ReqString(req), err)
+		// If err is hubmeta.ErrGVRNotRecognized, the reverse proxy will set the HTTP Status Code as 404.
+		return nil, err
 	}
-	if _, gvk = cm.restMapperManager.KindFor(gvr); gvk.Empty() {
-		return nil, hubmeta.ErrGVRNotRecognized
-	} else {
-		kind = gvk.Kind
+	listObj, err := generateEmptyListObjOfGVK(listGvk)
+	if err != nil {
+		klog.Errorf("failed to create ListObj for gvk %s for req: %s, %v", listGvk.String(), util.ReqString(req), err)
+		return nil, err
+	}
+	if err := setListObjSelfLink(listObj, req); err != nil {
+		klog.Errorf("failed to set selfLink for ListObj of gvk %s for req: %s, %v", listGvk.String(), util.ReqString(req), err)
+		return nil, err
 	}
 
-	// If the GVR information is recognized, return list or empty list
 	objs, err := cm.storage.List(key)
+	if err == storage.ErrStorageNotFound && proxyutil.IsListRequestWithNameFieldSelector(req) {
+		// When the request is a list request with FieldSelector "metadata.name", we should not return error
+		// when the specified resource is not found, to keep same as APIServer.
+		return listObj, nil
+	}
 	if err != nil {
-		if !errors.Is(err, storage.ErrStorageNotFound) {
-			return nil, err
-		} else if isPodKey(key) {
+		klog.Errorf("failed to list key %s for request %s, %v", key.Key(), util.ReqString(req), err)
+		return nil, err
+	}
+	if len(objs) == 0 {
+		if isKubeletPodRequest(req) {
 			// because at least there will be yurt-hub pod on the node.
 			// if no pods in cache, maybe all of pods have been deleted by accident,
 			// if empty object is returned, pods on node will be deleted by kubelet.
 			// in order to prevent the influence to business, return error here so pods
 			// will be kept on node.
+			klog.Warningf("get 0 pods for kubelet pod request, there should be at least one, hack the response with ErrStorageNotFound error")
+			return nil, storage.ErrStorageNotFound
+		}
+		// There's no obj to fill in the list, so just return.
+		return listObj, nil
+	}
```

Comment:
optimize Line184~205 as following will be more readable:
```
if err == storage.ErrStorageNotFound && proxyutil.IsListRequestWithNameFieldSelector(req) {
    // xxx
} else if err != nil {
    // xxx
} else if len(objs) == 0 {
    // xxx
}
```

Reply:
fixed